chore(clayui.com): adding the post of "Introducing Clay v3" (#2662)
liferay/clay: clayui.com/content/blog/2019-10-24-introducing-clay-v3.md

You can see our [quick start](/docs/get-started/how-to-use-clay.html) that will help you get started with Clay.

<div class="nav-toc">

-	[New Site](#new-site)
-	[React Components](#react-components)
	-	[New Components](#new-components)
	-	[Principles](#principles)
-	[Deprecation warning for Clay v2](#deprecation-warnings-for-clay-v2)
-	[Release Process](#release-process)
	-	[SemVer](#semver)
-	[Migrate Guide](#migrate-guide)
-	[3.x Roadmap](#3.x-roadmap)
	-	[Data Display](#data-display)
	-	[Data Visualization](#data-visualization)
	-	[Layout and ClayManager components](#layout-and-claymanager-components)
	-	[A11y component](#a11y-component)

</div>

### New Site

We introduced a visual redesign of our documentation and adding more specific content to the components.

We are taking the approach of merging the same CSS/Markup and React Component documents on the same page use tabs, this will make it easier for you to find more relevant information on the same subject. You can find only React documents or just CSS/Markup as well.

> All documents related to components in React and CSS, even satellites are all within the same `Components` session. We've put it all together to get content dispersed across the site and help you find documentation more easily.

Author: Oh yeah, sorry for not being clear about that 😅. Looks better now.

| [Popover](/docs/components/popover.html) | `@clayui/popover` |
| [Tabs](/docs/components/tabs.html) | `@clayui/tabs` |
| Slider | `@clayui/slider` |

#### Principles

With the framework change that impacts the way we use and think about the use of components, we developed these new components following React's [composition](https://reactjs.org/docs/composition-vs-inheritance.html) approach, you can read more about [Clay's fundamentals of composition in our documentation](/docs/foundations/composing.html) if you want to dig deeper.

### Deprecation warning for Clay v2

Since we are entering v2 and a major framework change for our components, we are starting to deprecated v2 components. This means that we will not be working on new features for v2 but will still continue to support bugs and security issues. We will detail in other posts about how the gradual depreciation of v2 will work.

### Release Process

We are adopting package independent release, releasing packages with their independent version, this will make it easier to make gradual and specific releases of each component. We will test how this release process will work.

#### SemVer

> We are adopting the [SemVer](https://semver.org/) versioning system to make it easy for you to upgrade our packages with confidence.

Author: Yeah that's true

```shell
MAJOR.MINOR.PATCH
```

> So you can see more often releases of `patch` and consequently` major` but that does not mean there will be big changes like v3 introducing a new rewrite. We hope these releases will be more continuous and help us iterator faster.

Reviewer: Not super sure what we are trying to communicate here. Can you give a bit more clarification on this part? One thing I might avoid is mention "major" changes since that can be scary to see lol

Author: haha yeah, in the end what I meant is that we want to use SemVer more in our favor, versions, beta, alpha ... and send the message that a "major" version is not necessarily a big change, like changing frameworks. I was trying to get the message across that we are not going to do this anytime soon but we are also going to release faster releases, so there might be a lot of "patch" versions for example.

Reviewer: Ah gotcha, maybe we could just say something along the lines of..

```suggestion
We hope that by following semver more strictly that it will give developers more confidence in what to expect with each release. Our goal is to provide confidence in new releases by providing as much backwards compatibility as possible and providing sufficient warnings to give users the best experience.
```

This way we don't need to mention any specific version itself and this way we can communicate our overall goal of providing consistency and confidence for users.

Author: Looks better I'll add this

### Migrate Guide

> We have prepared a migration guide from v2 to v3, keep in mind that v3 has had major changes this guide helps you come from Metal.js to React.js. We hope this can ease the pain for those unfamiliar with React.js.

Reviewer: Just removing some redundancy from what we have talked about above.
```suggestion
We have prepared a migration guide from v2 to v3. We hope this can ease the pain for those unfamiliar with React.js.
```

<div class="clay-ullist clay-ullist-margin-sm">

- [Migrate the Clay components from v2 to v3](/docs/migration/migrate-the-clay-components-from-v2-to-v3.html)

</div>

### 3.x Roadmap

> We are working on developer experience enhancements and delivering new features to accelerate development. We have a few things for our roadmap.

Reviewer: ```suggestion
We are working on developer experience enhancements and delivering new features. We have a few things for our roadmap.
```

accelerating development is sort of redundant of developer experience. I think it makes a bit more sense leaving it out.

Author: makes sense

#### Data Display

To help teams create complex and extensible data displays, Clay offers different Table components that can be used and configured on a high level.

> We will collect use cases and possible future data display cases, we will work on creating documentation and examples to help developers effectively display data across different applications.

Reviewer: ```suggestion
To help teams create complex and extensible data displays, Clay offers different Table components that can be used and configured on a high level. However, we are collecting current and future use cases so that we can create documentation and examples to help developers effectively display data across different applications. As we discover common patterns and uses, we will further evaluate the best way to provide components for users.
```

#### Data Visualization

> The proposal here is to analyze the problems we had with ClayCharts, many teams complained about its use, API, inconsistencies with Lexicon… the idea is to listen to all the people who worked with Charts to gather feedback and from that we can make a decision how let's work on this, whether it's a charts lib, a lib from scratch, API-only enhancements…

Reviewer: ```suggestion
The proposal here is to analyze the problems we had with ClayCharts, many teams complained about its use, API, and inconsistencies with Lexicon. The goal is to listen to current users who worked with Charts to gather feedback and evaluate the best direction forward, whether that be a new charts library or adding API enhancements to our current implementation.
```

> ClayCharts should be simple to create charts that Lexicon predicts and easy to customize to get out of specification.

Reviewer: ```suggestion
ClayCharts should allow for both simple use cases and also allow customization for further complexity.
```

An important point is performance, many onscreen charts should not be a problem, We will study the cases and help solve the performance issues.

The criteria to get there:

<div class="clay-ullist clay-ullist-margin-sm">

- Performance
- Flexibility
- Consistency with Lexicon
- Readable API

</div>

We are already working on [collecting feedbacks and you can leave your 2cents](https://github.com/liferay/clay/issues/2577).

#### Layout and ClayManager components

Still under research, we will be working to understand common component use cases that can become components in Clay that help develop Layouts more easily.

Since we have several contexts for different components it can be a bit annoying to add many nested contexts to your project's root application. We will work in an aggregating context that will bring all our contexts together in one place, making it easier for teams that use Clay as their base and not just some components.

#### A11y component

> A written component on top of react-axe that gives direct accessibility feedback on screen and only in the development environment. Inspired by the work for the new Facebook.

Reviewer: Maybe we should include a link to our storybook to show where people can see that integration.

Author: Hmm would be a good idea, but do you just say submit to the storybook or when we implement this new component?

Reviewer: oh wait, i think I misread this. Are you saying we are evaluating a new A11y component down the road?

Author: yes I added this in in the roadmap doc

Reviewer: ah gotcha. For that case then, I might actually leave it out since its fairly vague at the moment. If we want to include something on accessibility, we could just comment broadly how our goal is to continually increase accessibility of all of our components.

Author: Hm, it makes sense. We can post a post soon just talking about it.

> This can help reinforce our work in following accessibility guidelines and help product teams follow.

Reviewer: One last think we might want to add is a sort of conclusion to this post. It could simply be a short summary and thanks to those who have used v2 and expressing further desire to see v3 widely adopted and our willingness to help people in adoption. wdyt?

Author: Yeah that's great, good idea.

The criteria to get there:

<div class="clay-ullist clay-ullist-margin-sm">

- Performance
- Flexibility
- Consistency with Lexicon
- Readable API

</div>

We are already working on [collecting feedbacks and you can leave your 2cents](https://github.com/liferay/clay/issues/2577).

#### Layout and ClayManager components

Still under research, we will be working to understand common component use cases that can become components in Clay that help develop Layouts more easily.

Since we have several contexts for different components it can be a bit annoying to add many nested contexts to your project's root application. We will work in an aggregating context that will bring all our contexts together in one place, making it easier for teams that use Clay as their base and not just some components.

#### A11y component

A written component on top of react-axe that gives direct accessibility feedback on screen and only in the development environment. Inspired by the work for the new Facebook.

This can help reinforce our work in following accessibility guidelines and help product teams follow.

### Acknowledgments

> We are very excited about this delivery and are curious as to how people will use Clay going forward, we want to thank everyone who used Clay v2 and struggled to contribute, creating issues and sending PRs. Many thanks to all the comments we received about using Clay v2 that helped us to model this new version. Keeping this feedback is very important so we can keep evolving our implementation and we are available to hear and understand your use cases. We hope adoption continues to grow.

Reviewer: Just a few suggestions.

```suggestion
We are very excited about this release and are looking forward to see how people use Clay going forward, we want to thank everyone who used Clay v2 and who contributed by creating issues and sending PRs. Many thanks to all the comments we received about using Clay v2 that helped us to model this new version. Keeping this feedback is very important so that we can keep evolving and creating a great experience for end users and developers. We look forward to continue growth.
```
